### PR TITLE
[FLINK-29185][hive] fix class not found exception for creating temporary function using jar with Hive dialect

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To fix class not found exception for creating temporary function using jar with Hive dialect

## Brief change log
  -  If it's for creating temporary function, first try to register the resources.

## Verifying this change
Modified test in `HiveDialectITCase#testCreateFunctionUsingJar`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 1]
```diff
@@ -745,32 +745,40 @@ public void testTemporaryFunctionUDAF() throws Exception {
...
         assertThat(
                         CollectionUtil.iteratorToList(
                                         tableEnv.executeSql("select add_one(x) from src").collect())
                                 .toString())
                 .isEqualTo("[+I[2], +I[3]]");
+
         // test create temporary function using jar
+        // create a new jarfile with a new class name
+        udfClass = "addOne1";
+        udfCode = String.format(udfCodeTemplate, udfClass);
```
Comment (wuchong) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:774]: It seems there is a flaw in the current design/implementation that the previous user jar is not removed from the classloader which may lead to class conflict when using the same class name but a different code. For example, if the code of `addOne1` is `content - 1`, the executed result might still be `content + 1`. @lsyldliu could you create an issue to track that?

**Review by wuchong — APPROVED** [review]
LGTM.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java (hunk 1 above)
Comment (lsyldliu) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:774]: IMO, we don't support remove jar, the class conflict is an expected behavior if the class name is same?

Comment (wuchong) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:774]: But the query only references one user jar. But the job contains 2 jars.

**Pull request merged**